File: Archived/11단계/2798.java
import java.util.*;
import java.io.*;

public class Main{
    static BufferedReader br;
    static BufferedWriter bw;
    static int N,M,max;
    static int[] cards;
    public static void main(String[] args) throws IOException{
        br = new BufferedReader(new InputStreamReader(System.in));
        bw = new BufferedWriter(new OutputStreamWriter(System.out));
        String[] tmp = br.readLine().split(" ");
        N = Integer.parseInt(tmp[0]);
        M = Integer.parseInt(tmp[1]);
        cards = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        recur(-1,0,0);
        bw.write(String.valueOf(max)+'\n');
        bw.flush();bw.close();br.close();

    }
    static void recur(int idx,int sum, int cnt){
        if(sum>M)
            return;

        if(cnt ==3) {
            if(sum>max) {
                max = sum;
            }
            return;
        }
        
        for(int i=idx+1;i<N;i++)
            recur(i,sum+cards[i],cnt+1);

    }
}

File: Archived/11단계/2798.py
import sys

N, M = map(int,sys.stdin.readline().strip().split())
cards = list(map(int,sys.stdin.readline().strip().split()))

closest = -1
for i in range(N-2):
    a = cards[i]
    for j in range(i+1,N-1):
        b = cards[j]
        for k in range(j+1,N):
            c = cards[k]
            tmp = a + b + c
            if tmp <= M and tmp > closest:
                closest = tmp
print(closest)
